model/llama/model.go

```diff
@@ -52,38 +57,177 @@ type SelfAttention struct {
 	Output *nn.Linear `ggml:"attn_output"`
 }
 
-func (sa *SelfAttention) Forward(ctx ml.Context, hiddenState, positionIDs ml.Tensor, cache model.Cache, opts *Options) ml.Tensor {
-	batchSize := hiddenState.Dim(0)
-	headDim := opts.hiddenSize / opts.numHeads
-
-	q := sa.Query.Forward(ctx, hiddenState)
-	q = q.Reshape(ctx, headDim, opts.numHeads, batchSize)
-	// q = q.Rope(ctx, positionIDs, opts.RopeFactors, opts.ropeDim, opts.ropeBase, opts.ropeScale)
-
-	k := sa.Key.Forward(ctx, hiddenState)
-	k = k.Reshape(ctx, headDim, opts.numKVHeads, batchSize)
-	// k = k.Rope(ctx, positionIDs, opts.RopeFactors, opts.ropeDim, opts.ropeBase, opts.ropeScale)
-
-	v := sa.Value.Forward(ctx, hiddenState)
-	v = v.Reshape(ctx, headDim, opts.numKVHeads, batchSize)
-
-	k, v = cache.Put(ctx, k, v, cache.Options)
-
-	q = q.Permute(ctx, 0, 2, 1, 3).Contiguous(ctx)
-	k = k.Permute(ctx, 0, 2, 1, 3).Contiguous(ctx)
-	v = v.Permute(ctx, 1, 2, 0, 3).Contiguous(ctx)
+func (sa *SelfAttention) Forward(ctx ml.Context, hiddenState ml.Tensor, offset int32, cache model.Cache, opts *Options) ml.Tensor {
+	// Ref: https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/llama.py
+	shape := hiddenState.Shape()
+	B := shape[0]
+	L := shape[1]
+	n_heads := opts.numHeads
+	n_kv_heads := opts.numKVHeads
+
+	head_dim := opts.hiddenSize / opts.numHeads
+	scale := math.Pow(float64(head_dim), -0.5)
+
+	queries := sa.Query.Forward(ctx, hiddenState)
+	keys := sa.Key.Forward(ctx, hiddenState)
+	values := sa.Value.Forward(ctx, hiddenState)
+
+	queries = queries.Reshape(ctx, B, L, n_heads, -1).Permute(ctx, 0, 2, 1, 3)
+	keys = keys.Reshape(ctx, B, L, n_kv_heads, -1).Permute(ctx, 0, 2, 1, 3)
+	values = values.Reshape(ctx, B, L, n_kv_heads, -1).Permute(ctx, 0, 2, 1, 3)
+
+	queries = sa.Rope(ctx, queries, offset, opts)
+	keys = sa.Rope(ctx, keys, offset, opts)
+
+	// TODO - The current cache algorithm doesn't work properly on MLX due to stride values being onconsistent with
+	// the GGML backend.  When this caching support comes back, the input should be truncated to just the latest token
+	// keys, values = cache.Put(ctx, keys, values, cache.Options)
+
+	// TODO - Some backends will support various "fast" variations of common operations, but not all will
+	// Either we need to build the "slow" version in the backend, or the model has to be able to discover
+	// if the backend supports the fast operation and if not fall back to the slow version.
+	// GGML does not have a fast scaled dot product attention, so this block is currently conditional
+	var output ml.Tensor
+	if opts.GGML { // TODO - should be based on an abstracted lookup for "fast scaled dot product attention" support
```

Comment:
You can create an interface `FastScaledDotProductAttention` and fallback to manual sdpa if the context doesn't implement it. something like

```go
if sdpa, ok := ctx.(ml.FastScaledDotProductAttention); ok {
  hiddenState = sdpa.Forward(...)
} else {
  // manual sdpa
}
```